[Hold] Handle illegal characters in windows filenames

abought opened the pull request:
Refs [#OFF-7]
## Purpose

It is possible for a mac user to upload a file with special characters that would make it impossible to download that file on windows; the user would see errors on every sync attempt.

This PR introduces a new "alias" feature that downloads the file to a unique filename (special characters replaced with underscores on windows), and allows it to be updated on the OSF despite having a different name locally.
## Scope/ limitations

Deals with windows special characters, but it does not deal with length limitations or other such errors on windows. A 260-char pathname will still be illegal no matter what.

Only changes behavior on windows. On mac, filenames are allowed to exist as is.

Only changes paths for files and folders. Will not handle aliasing of project names.
## Summary of changes
- Change how paths in filenames are constructed for `LocalOperationThing`s. (Entirely possible I missed some cases; let me know)
- Change auditor so that local and remote paths can be matched up, even when the local path is aliased
- Introduce a function that creates a platform-specific filename alias
- Add a new `alias` field to DB files/folders
- Add various accessors for both the aliased path, and the unmodified path (need both to successfully audit and match local/remote paths)
## Requirements/ test criteria

A sample project structure used to test operations is here (it can be forked for convenience): https://test.osf.io/cq3h7/

This can be tested on mac by editing the two `sys.platform` checks in `utils/__init__.py` to be `if True` blocks instead. On mac, it will not alias filenames.
1. [x] A filename created or renamed on the local computer should always be uploaded with exactly that name. (trust the filesystem to ensure the name is legal)
   - [x] And the alias should be cleared from the database for that file when the rename/move operation is complete
2. [x] A filename downloaded on mac should always have exactly the same name as it does on the OSF. No transformations should apply, and the “alias” column should always be empty for DB entries
3. [x] If a file with special characters in the name is uploaded to the OSF, it should be aliased when stored locally (the special characters in the windows filename should be replaced with underscores)
4. [x] All CRUD operations should operate seamlessly for a file whose name has not been aliased
   - On Mac 
   - On Windows
5. [x] All CRUD operations should work correctly for a file whose name is aliased (on windows) due to special characters.
6. [x] When an aliased file is modified, the change should affect the original file on the OSF seamlessly (even though local and remote name are slightly different). It should not upload a new file with the aliased local name.
7. [x] In this version, aliasing should work for files and folders, but not projects. It should work in subfolders as well as at the top level.
   - [x] If both a folder and a file require aliasing, this should work seamlessly.
8. [x] If a file is aliased, and then renamed on the OSF, the file should correctly be renamed locally. If the new filename does not require an alias, the alias field of the DB should be wiped.
   - [x] If a file has a good local name, but is then renamed on the OSF (to something requiring an alias), the rename should propagate locally.
## TODO
- [x] Creating a new file (normal name) in an existing folder (with an aliased name) results in the file being created in the wrong place. On the remote server, the new file ends up in the top-level project instead.  *Works on Mac OS (RemoteCreateFile + DatabaseCreateFile operations triggered by watchdog)
  - But the new version may get uploaded to the correct place later during an audit cycle.
- [ ] Editing an existing image file gives "temp file not found" errors on windows. (this may be unrelated to this PR)
- [ ] On windows, watchdog sometimes sends two creation events for a file; the second create--> update event sometimes uploads files to the wrong places.

abought committed:
Add alias field to handle filenames with illegal chars

Remove osf_path field, because this was overridden by a property, and never really part of the schema anyway

abought committed:
Utility function for safe filename alias on a given platform

abought committed:
Add alias field to operation context; it will be undefined if remote is undefined

abought committed:
Quick rearrange of methods

abought committed:
Move repr should show dest

abought committed:
Start connecting alias to operations

- OperationContext gains alias property (used by database ops) and safe_name property (used by local file ops)
- DB model gains safe_name property as well (to distinguish from alias)

abought committed:
Fix local fn to db func

abought committed:
Merge branch 'develop' of https://github.com/CenterForOpenScience/OSF-Offline into feature/off7-windows-filenames

abought committed:
Add docstring to aid future debugging

samchrisinger commented on osfoffline/utils/__init__.py:
```diff
@@ -37,6 +39,38 @@ def hash_file(path, *, chunk_size=65536):
+    if sys.platform in ('win32', 'cygwin'):
+        # https://support.microsoft.com/en-us/kb/177506
+        illegal_chars = r'\/:*?"<>|'
+        new_fn = re.sub(r'[{}]'.format(illegal_chars), '_', basename)
```
Might consider compiling this and storing it in a variable outside of this function, i.e.:
`ILLEGAL_CHARS = re.compile(r'[\/:*?"<>|]')

samchrisinger commented on osfoffline/utils/__init__.py:
```diff
@@ -37,6 +39,38 @@ def hash_file(path, *, chunk_size=65536):
+        n = 1
+        while parent:
+            # If parent node is provided, loop through until a valid filename (not in use) is available
+            if Session().query(models.File.name).filter(models.File.alias == new_fn,
```
Could be better to create a set of taken filenames outside of this loop rather than re-running the query at each iteration.

abought replied:
I'll make this change, and be happier for it. Thanks for the nudge. :)

abought replied:
Oh, wait, now I remember- `new_fn` is redefined inside the loop. On each pass, it's asking if a different filename exists.

(If no results are found, `all` returns an empty list, but `one` would throw an exception)

abought committed:
Fix flake warnings

abought committed:
Fix alias bug
(self.db was violating causality)

abought committed:
Define illegal chars as compiled re outside function

abought committed:
Fix regex

abought committed:
Add notes on function inputs

abought committed:
Annotate types on docstring

abought committed:
Tweak auditor to handle aliased paths

Auditor local/remote/db maps are keyed on path, so aliased paths break the auditor. This change allows the DB object to be found even when the local file is an alias of the db/remote item, and changes the local map to key on unaliased path.

(this way, comparing local and remote map keys should not generate spurious move events)

abought committed:
More informative repr

abought committed:
Filter alias entries out of changes and satisfy getattr usage

abought committed:
Correctly handle unaliased file in aliased parent folder

Fix an error that caused file deletion/data loss in certain cases.

Test data in https://test.osf.io/cq3h7/

abought committed:
Fix filename helper.

 Duplicate aliases should append a single number, not (1)(2)(3)...

abought committed:
Fix LocalMove to handle aliased paths

abought committed:
Appease the flake gods

abought committed:
Fix LocalMoves

 OperationContext was constructing a non-safe path, so it needs to be correctly converted into a platform-safe alias. Rather than edit the broadly used context, added helper to get the matching DB unsafe path and used helper in local move to break down the fake path and rebuild with an alias.

abought commented:
Working-ish, but found a few new edge cases while testing.

abought committed:
Add a comment to function

abought commented:
Update: creating a new file in a folder with an aliased name works on Mac OS, but not Windows.

Procedure was to copy-paste an existing file (using finder on Mac and explorer on windows) into the same folder. On Mac this fired a pure create operation. On Windows...

```
[DEBUG][2016-01-20 14:50:56,464][Thread-5][osfoffline.sync.ext.watchdog]: Watchdog event fired: <DirModifiedEvent: src_path='D:\\Users\\cosadmin\\Desktop\\OSF\\Windows2 Fork of Aliased filename tester - 89d6r\\OSF Storage'>
[DEBUG][2016-01-20 14:50:59,057][Thread-5][osfoffline.sync.ext.watchdog]: Watchdog event fired: <FileCreatedEvent: src_path='D:\\Users\\cosadmin\\Desktop\\OSF\\Windows2 Fork of Aliased filename tester - 89d6r\\OSF Storage\\cats _ dogs Yes\\osf_test_file3 - Copy - Copy.jpg'>
[DEBUG][2016-01-20 14:50:59,071][Thread-5][osfoffline.sync.ext.watchdog]: Watchdog event fired: <FileModifiedEvent: src_path='D:\\Users\\cosadmin\\Desktop\\OSF\\Windows2 Fork of Aliased filename tester - 89d6r\\OSF Storage\\cats _ dogs Yes\\osf_test_file3 - Copy - Copy.jpg'>


[DEBUG][2016-01-20 14:50:59,080][Thread-5][osfoffline.sync.ext.watchdog]: Watchdog event fired: <DirModifiedEvent: src_path='D:\\Users\\cosadmin\\Desktop\\OSF\\Windows2 Fork of Aliased filename tester - 89d6r\\OSF Storage\\cats _ dogs Yes'>
[DEBUG][2016-01-20 14:51:01,079][Thread-21][osfoffline.sync.ext.watchdog]: Watchdog event dispatched: <FileModifiedEvent: src_path='D:\\Users\\cosadmin\\Desktop\\OSF\\Windows2 Fork of Aliased filename tester - 89d6r\\OSF Storage\\cats _ dogs Yes\\osf_test_file3 - Copy - Copy.jpg'>
[INFO][2016-01-20 14:51:01,079][Thread-21][osfoffline.sync.local]: Modification event  for file: D:\Users\cosadmin\Desktop\OSF\Windows2 Fork of Aliased filename tester - 89d6r\OSF Storage\cats _ dogs Yes\osf_test_file3 - Copy - Copy.jpg
[INFO][2016-01-20 14:51:01,096][Thread-21][osfoffline.sync.local]: Creation event for file: D:\Users\cosadmin\Desktop\OSF\Windows2 Fork of Aliased filename tester - 89d6r\OSF Storage\cats _ dogs Yes\osf_test_file3 - Copy - Copy.jpg
[DEBUG][2016-01-20 14:51:01,108][Thread-21][osfoffline.sync.ext.watchdog]: Watchdog event dispatched: <FileCreatedEvent: src_path='D:\\Users\\cosadmin\\Desktop\\OSF\\Windows2 Fork of Aliased filename tester - 89d6r\\OSF Storage\\cats _ dogs Yes\\osf_test_file3 - Copy - Copy.jpg'>
[INFO][2016-01-20 14:51:01,109][Thread-2][osfoffline.tasks.operations]: Starting <RemoteCreateFile(<OperationContext(node=<Node(89d6r, title=Windows2 Fork of Aliased filename tester, path=D:\Users\cosadmin\Desktop\OSF\Windows2 Fork of Aliased filename tester - 89d6r)>, local=D:\Users\cosadmin\Desktop\OSF\Windows2 Fork of Aliased filename tester - 89d6r\OSF Storage\cats _ dogs Yes\osf_test_file3 - Copy - Copy.jpg, db=None, remote=None)>)>
[INFO][2016-01-20 14:51:01,109][Thread-21][osfoffline.sync.local]: Creation event for file: D:\Users\cosadmin\Desktop\OSF\Windows2 Fork of Aliased filename tester - 89d6r\OSF Storage\cats _ dogs Yes\osf_test_file3 - Copy - Copy.jpg
[INFO][2016-01-20 14:51:01,128][Thread-2][requests.packages.urllib3.connectionpool]: Resetting dropped connection: test-files.osf.io
[DEBUG][2016-01-20 14:51:02,633][Thread-2][requests.packages.urllib3.connectionpool]: "PUT /v1/resources/89d6r/providers/osfstorage/569fdf5bda3e240181422ecc/?name=osf_test_file3+-+Copy+-+Copy.jpg HTTP/1.1" 201 1089
[INFO][2016-01-20 14:51:02,636][Thread-2][osfoffline.tasks.operations]: Starting <DatabaseCreateFile(<OperationContext(node=<Node(89d6r, title=Windows2 Fork of Aliased filename tester, path=D:\Users\cosadmin\Desktop\OSF\Windows2 Fork of Aliased filename tester - 89d6r)>, local=None, db=None, remote=<RemoteFile(569fe526da3e240181422ed6, osf_test_file3 - Copy - Copy.jpg, file, 569fdf5bda3e240181422ecc>)>)>
[INFO][2016-01-20 14:51:02,665][Thread-2][osfoffline.tasks.notifications]: Notification: <Event (INFO) Create Remote File: D:\Users\cosadmin\Desktop\OSF\Windows2 Fork of Aliased filename tester - 89d6r\OSF Storage\cats _ dogs Yes\osf_test_file3 - Copy - Copy.jpg>
[INFO][2016-01-20 14:51:02,669][Thread-2][osfoffline.tasks.operations]: Starting <RemoteCreateFile(<OperationContext(node=<Node(89d6r, title=Windows2 Fork of Aliased filename tester, path=D:\Users\cosadmin\Desktop\OSF\Windows2 Fork of Aliased filename tester - 89d6r)>, local=D:\Users\cosadmin\Desktop\OSF\Windows2 Fork of Aliased filename tester - 89d6r\OSF Storage\cats _ dogs Yes\osf_test_file3 - Copy - Copy.jpg, db=None, remote=None)>)>
[DEBUG][2016-01-20 14:51:02,680][Thread-2][osfoffline.tasks.operations]: File already exists; will run update operation instead
[INFO][2016-01-20 14:51:02,680][Thread-2][osfoffline.tasks.operations]: Starting <RemoteUpdateFile(<OperationContext(node=<Node(89d6r, title=Windows2 Fork of Aliased filename tester, path=D:\Users\cosadmin\Desktop\OSF\Windows2 Fork of Aliased filename tester - 89d6r)>, local=D:\Users\cosadmin\Desktop\OSF\Windows2 Fork of Aliased filename tester - 89d6r\OSF Storage\cats _ dogs Yes\osf_test_file3 - Copy - Copy.jpg, db=<DBFile(569fe526da3e240181422ed6, osf_test_file3 - Copy - Copy.jpg, file, 569fdf5bda3e240181422ecc>, remote=None)>)>
[DEBUG][2016-01-20 14:51:03,771][Thread-2][requests.packages.urllib3.connectionpool]: "PUT /v1/resources/89d6r/providers/osfstorage/?name=osf_test_file3+-+Copy+-+Copy.jpg HTTP/1.1" 201 1073
[INFO][2016-01-20 14:51:03,772][Thread-2][osfoffline.tasks.operations]: Starting <DatabaseUpdateFile(<OperationContext(node=<Node(89d6r, title=Windows2 Fork of Aliased filename tester, path=D:\Users\cosadmin\Desktop\OSF\Windows2 Fork of Aliased filename tester - 89d6r)>, local=None, db=<DBFile(569fe526da3e240181422ed6, osf_test_file3 - Copy - Copy.jpg, file, 569fdf5bda3e240181422ecc>, remote=<RemoteFile(569fe527da3e240181422ed8, osf_test_file3 - Copy - Copy.jpg, file, 569fdf5bda3e240181422ecc>)>)>
[INFO][2016-01-20 14:51:03,782][Thread-2][osfoffline.tasks.notifications]: Notification: <Event (INFO) Update Remote File: D:\Users\cosadmin\Desktop\OSF\Windows2 Fork of Aliased filename tester - 89d6r\OSF Storage\cats _ dogs Yes\osf_test_file3 - Copy - Copy.jpg>

```

abought commented:
It seems like the update operation is actually acting as a create operation. 

There are two core issues:
1. It appears that d1e6b9eeb005d072c27d3bcb4b88a31783a41845  allowed updates to respond to 201 as well as 200.

2. When a create is turned into an update, there appears to be an edge case in which the parent is not set correctly, leading to this bug.

Create URL:
        url = '{}/v1/resources/{}/providers/{}/{}'.format(settings.FILE_BASE, self.node.id, parent.provider, parent.osf_path)
[DEBUG][2016-01-20 14:57:08,136][Thread-2][requests.packages.urllib3.connectionpool]: "PUT /v1/resources/89d6r/providers/osfstorage/569fdf5bda3e240181422ecc/?name=osf_test_file3+-+Copy+-+Copy+-+Copy.jpg HTTP/1.1" 201 1103


Update URL:
        url = '{}/v1/resources/{}/providers/{}/{}'.format(settings.FILE_BASE, self.node.id, self.db.provider, self.db.osf_path)
[DEBUG][2016-01-20 14:57:09,187][Thread-2][requests.packages.urllib3.connectionpool]: "PUT /v1/resources/89d6r/providers/osfstorage/?name=osf_test_file3+-+Copy+-+Copy+-+Copy.jpg HTTP/1.1" 201 1087

samchrisinger commented on osfoffline/sync/ext/auditor.py:
```diff
@@ -273,26 +318,38 @@ def _collect_node_local(self, root, acc, db_map):
-                created.add(i)
-                deleted.add(i)
+        for k in keys_in_both:
+            if source[k].fid != target[k].fid and target[k].is_alias:
```
Aren't aliased keys being filtered out of `source` and `target` above?

samchrisinger replied:
For posterity: We really need a clearer understanding of the consequences of this. My understanding is that `keys_in_both` will contain no elements which are aliased in source or target. Consequently this branch would never be run.

samchrisinger commented:
Is it possible for a Node name to contain illegal characters?

samchrisinger commented:
resolved ^

samchrisinger commented:
Talking with Andy, this PR really needs to be held up until some unit tests are in place. The ideal strategy:

1. using a branch based off of develop, write tests to formalize the behavior of _diff
2. merge this PR into that branch
3. Make sure the tests are not broken by this PR
